contracts/token/BiconomyTokenPaymaster.sol

```diff
@@ -218,19 +195,76 @@ contract BiconomyTokenPaymaster is BasePaymaster, ReentrancyGuard, TokenPaymaste
-        (bool success, bytes memory returndata) = address(oracleAggregator).staticcall(_data);
-        exchangeRate = 0;
+        (bool success, bytes memory returndata) = address(_oracleAggregator).staticcall(_data);
+        exchangeRate = 0; // this is assigned for fallback
         if(success) {
             exchangeRate = abi.decode(returndata, (uint256));
         }
     }
 
+    
+    // review could pack attributes in single / 2 structs. A label field string memory label can be added
+    /**
+     * @dev Helper function to trigger periodic swap to convert tokens received inside this paymaster back to native tokens and immediately deposit on entry point. 
+     * @notice actual swap data is generated offchain basen on i. router ii. token to swap iii. amount of tokens iv. route v. slippage etc
+     * @notice Biconomy may not use feeReceiver to be paymaster contract itself. In this case tokens do not need to be pulled or swapped from the contract
+     * @param _dexRouter dex router/adapter address
+     * @param _swapData calldata for making the swap on chosen router. as function signature also depends on the choice
+     * @param _approveRouter caller sends if the router being used is to be approved or not
+     * @param _token ERC20 token address being swapped. useless is above bool flag is false
+     * @param _amount ERC20 token amount to be approved. useless is above bool flag is false
+     * @param _maxDepositToEP sender can pass maximum amount of gas to be deposited to entry point 
+     */
+    function swapTokenForETHAndDeposit(address _dexRouter, bytes calldata _swapData, bool _approveRouter, address _token, uint256 _amount, uint256 _maxDepositToEP) public nonReentrant onlyOwner returns(bool success, uint256 depositAmount) {
+        // only proceed if adapter is not 0 address
+        require(_dexRouter != address(0), "Router can not be zero address");
+        // make approval to router 
+        if(_approveRouter) {
+        SafeTransferLib.safeApprove(_token, _dexRouter, _amount);
+        }
+
+        // make the swap
+        // review could take snapshot of token balance and eth balance before and after the swap
+         (bool success, bytes memory returndata) = address(_dexRouter).call(_swapData);
+
+        uint256 depositAmount = address(this).balance; 
+
+        if(depositAmount > _maxDepositToEP) {
+            depositAmount = _maxDepositToEP;
+        }
+
+        if(depositAmount > 0) {
+        // entrypoint.depositTo
+        IEntryPoint(entryPoint).depositTo{value: depositAmount}(address(this));
+        }
+
+        emit TokenSwappedAndGasDeposited(_dexRouter, _token, _swapData, success, depositAmount);
+    }
+
+    // method to pull any excess eth in the paymaster contract 
+    function withdrawAllETHTo(address dest) public nonReentrant {
+        require(owner() == msg.sender, "only owner can withdraw"); // add revert code
```

Comment:
At some places onlyOwner modifier is used, and here owner() == msg.sender is used. We should use one way of adding owner restrictions.

Reply:
keeping onlyOwner. done